src/main/java/org/jetbrains/kotlinx/lincheck/runner/ParallelThreadsExecutor.kt

```diff
@@ -0,0 +1,196 @@
+/*-
+ * #%L
+ * Lincheck
+ * %%
+ * Copyright (C) 2019 - 2020 JetBrains s.r.o.
+ * %%
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ * 
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Lesser Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Lesser Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/lgpl-3.0.html>.
+ * #L%
+ */
+package org.jetbrains.kotlinx.lincheck.runner
+
+import kotlinx.atomicfu.atomicArrayOfNulls
+import org.jetbrains.kotlinx.lincheck.runner.ParallelThreadsRunner.TestThread
+import java.util.concurrent.ExecutionException
+import java.util.concurrent.TimeoutException
+import java.util.concurrent.locks.LockSupport
+import kotlin.math.min
+
+/**
+ * This class executes a set of test thread executions in an internal thread pool.
+ */
+class ParallelThreadsExecutor(private val nThreads: Int, runnerHash: Int) {
+    // null, a waiting TestThread, TestThreadExecution or SHUTDOWN for each thread
+    private val tasks = atomicArrayOfNulls<Any>(nThreads)
+    // null, a waiting submitting Thread, DONE or an exception for each thread
+    private val results = atomicArrayOfNulls<Any>(nThreads)
+    // how many iterations threads spend in active wait until parking
+    private var spinCount = 40000
+    // for adaptive spin count changing
+    @Volatile
+    private var wasParked: Boolean = false
+    private var wasParkedBalance: Int = 0 // to negate dispersion because of i.e. verifier
+
+    init {
+        repeat(nThreads) { iThread ->
+            TestThread(iThread, runnerHash, testThreadRoutine(iThread)).start()
+        }
+    }
+
+    /**
+     * Submits a set of test thread executions to the thread pool and waits until all executions are completed.
+     * The number of executions should be equal to [nThreads].
+     * @throws TimeoutException if more than [timeoutMs] passed
+     * @throws ExecutionException if an unexpected exception happened
+     */
+    fun submitAndAwaitExecutions(executions: Array<TestThreadExecution>, timeoutMs: Long? = null) {
```

Comment:
We do not use the fact that `TestThreadExecution`-s are submitted, let's submit `Runnable`-s